File: src/eu/yeger/gramofo/fol/graph/ModelChecker.java
package eu.yeger.gramofo.fol.graph;

import eu.yeger.gramofo.fol.FOLParser;
import eu.yeger.gramofo.fol.Settings;
import eu.yeger.gramofo.fol.formula.*;

import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Set;


/**
 * This class is used to check if the drawn graph is a model of the formula. Both are available throw the datamodel.
 * (please be patient I wrote this drunk =). I think its a little bit necessary for this theory stuff)
 */
public class ModelChecker {

    private Graph graph;
    private FOLFormulaHead formulaHead;
    private HashMap<String, Set<Vertex>> oneArySymbolTable;
    private HashMap<String, Set<Edge>> twoArySymbolTable;
    private HashMap<String, String> symbolTypeTable;
    private HashMap<String, Vertex> bindVariableValues;
    private HashSet<String> infixPredicates;



    /**
     * Checks if the graph, which is drawn, is a model from the formula.
     * @param graph the interpretation as a graph.
     * @param formulaHead the formula to be checked
     * @return null if the graph is model of the formula and a String containing a error message else.
     */
    public String checkIfGraphIsModelFromFormula(Graph graph, FOLFormulaHead formulaHead) {
        this.graph = graph;
        this.formulaHead = formulaHead;
        this.oneArySymbolTable  = new HashMap<>();
        this.twoArySymbolTable  = new HashMap<>();
        this.symbolTypeTable    = new HashMap<>();
        this.bindVariableValues = new HashMap<>();
        this.infixPredicates    = new HashSet<>( Arrays.asList(new FOLParser().getSettings().getSetting(Settings.INFIX_PRED)) );

        try {
            checkIfInputIsValid();
            loadGraphSymbols();
            loadFormulaSymbols();
            checkTotality();

            if (checkModel(formulaHead.getFormula())) {
                return null;
            } else {
                return "";
            }
        } catch (ModelCheckException mce) {
            return mce.getMessage();
        }
    }

    public boolean checkModel(Graph graph, FOLFormulaHead formulaHead) {
        String result = checkIfGraphIsModelFromFormula(
                graph,
                formulaHead
        );

        return result == null || !result.equals("");
    }


    private void checkIfInputIsValid() throws ModelCheckException {
        if(graph == null) {
            throw new ModelCheckException("[ModelChecker][Internal error] graph is null.");
        }
        if (formulaHead == null) {
            throw new ModelCheckException("You must parse a formula first");
        }

        if (graph.getVertexes().size() < 1) {
            throw new ModelCheckException("The graph must contain at least one element. This is because the vertexes represent " +
                "the universe and a universe must contain at least one element.");
        }
    }


    /**
     * Iterates over the graph and puts all found symbols in the symbol tables.
     * @throws ModelCheckException if a symbol is used with different arities within the graph and
     * if a 1-ary function is not right-unique
     */
    private void loadGraphSymbols() throws ModelCheckException {
        graph.getVertexes().forEach( vertex ->
            vertex.getStringAttachments().forEach(symbol -> {
                String symbolType = Character.isUpperCase(symbol.charAt(0)) ? "P-1" : "F-0";
                symbolTypeTable.put(symbol, symbolType);
                Set<Vertex> relationSet = oneArySymbolTable.getOrDefault(symbol, new HashSet<>());
                if( symbolType.equals("F-0") && relationSet.size() != 0) {
                    throw new ModelCheckException("The 0-ary function symbol '" + symbol + "' can only be assigned to one vertex.");
                }
                relationSet.add(vertex);
                oneArySymbolTable.put(symbol,relationSet);
            })
        );

        graph.getEdges().forEach(edge ->
            edge.getStringAttachments().forEach(symbol -> {
                String symbolType = Character.isUpperCase(symbol.charAt(0)) || infixPredicates.contains(symbol) ? "P-2" : "F-1";
                symbolTypeTable.putIfAbsent(symbol, symbolType);
                if (!symbolType.equals(symbolTypeTable.get(symbol))) {
                    throw new ModelCheckException("The symbol '" + symbol + "' is defined with different arities within the graph.");
                }
                Set<Edge> relationSet = twoArySymbolTable.getOrDefault(symbol, new HashSet<>());
                if( symbolType.equals("F-1") && relationSet.stream().anyMatch( otherEdge -> edge.getFromVertex() == otherEdge.getFromVertex() )) {
                    throw new ModelCheckException("The 1-ary function '" + symbol + "' has at least for one vertex two function values. A function must be right-unique.");
                }
                relationSet.add(edge);
                twoArySymbolTable.put(symbol,relationSet);
            })
        );
    }

    /**
     * Iterates over the formula and add new symbols to the symbol tables.
     * @throws ModelCheckException if there are symbols used with different meanings
     */
    private void loadFormulaSymbols() throws ModelCheckException {
        formulaHead.getSymbolTable().forEach( (symbol, type) -> {
            symbolTypeTable.putIfAbsent(symbol, type);
            if (type.equals("P-1") || type.equals("F-0")) {
               oneArySymbolTable.putIfAbsent(symbol, new HashSet<>());
            } else if (type.equals("P-2") || type.equals("F-1")) {
               twoArySymbolTable.putIfAbsent(symbol, new HashSet<>());
            }
            String typeInGraph = symbolTypeTable.get(symbol);
            if (!type.equals(typeInGraph)) { // types are different?
                if (type.equals("V")) {
                    throw new ModelCheckException("The symbol '" + symbol + "' is defined in the formula as a bound variable " +
                        "but in the graph it is a function symbol. You cannot use one symbol twice for different use cases.");
                } else {
                    throw new ModelCheckException("The arity of the symbol '" + symbol + "' in the graph differ from the arity used in the formula.");
                }
            }
        });
    }


    /**
     * Functions mus be left total. Therefor this method checks all function symbols, if they are defined for all inputs.
     * @throws ModelCheckException if some function symbols aren't defined for all inputs.
     */
    private void checkTotality() throws ModelCheckException {
        symbolTypeTable.forEach((symbol, type) -> {
            switch (type) {
                case "F-0":
                    if (oneArySymbolTable.get(symbol).size() != 1) {
                        throw new ModelCheckException("The 0-ary function '" + symbol + "' must be defined. Please add it to the graph.");
                    }
                    break;
                case "F-1":
                    Set<Edge> relationSet = twoArySymbolTable.get(symbol);
                    graph.getVertexes().forEach(vertex -> {
                        if (relationSet.stream().noneMatch(edge -> edge.getFromVertex() == vertex)) {
                            throw new ModelCheckException("The 1-ary function '" + symbol + "' must be total. Please be sure that it is defined for all vertexes.");
                        }
                    });
            }
        });
    }


    /**
     * This function does the main job: the model checking. It handles the top-level type of the given formula
     * and call it self recursively on all children.<br>
     * The base case are: <br>
     * - the bind variables, which are associated with an real element throw an quantifier before<br>
     * - 0-are function symbols, which must be specified throw the graph <br>
     * Predicates and function interpretation are given throw the graph too.<br>
     * The Equality is simply implemented as the equality of two vertexes: vertexFromChild1 == vertexFromChild2 ?.
     * This is possible, because every term can be interpreted with a vertex.
     * @throws ModelCheckException if the data is invalid. This should not happen.
     */
    private boolean checkModel(FOLFormula formula) throws ModelCheckException {
        //note: this could also be done with inheritance. This would maybe the cleaner solution but I did not want to mix this could wit the datamodel.
        //Therefor I decide to make a switch case
        switch (formula.getType()) {
            case Forall:
                return graph.getVertexes().stream().allMatch(vertex -> {
                   bindVariableValues.put(formula.getChildAt(0).getName(), vertex);
                   return checkModel(formula.getChildAt(1));
                });

            case Exists:
                return graph.getVertexes().stream().anyMatch(vertex -> {
                    bindVariableValues.put(formula.getChildAt(0).getName(), vertex);
                    return checkModel(formula.getChildAt(1));
                });

            case Not:
                return !checkModel(formula.getChildAt(0));

            case And:
                return checkModel(formula.getChildAt(0)) && checkModel(formula.getChildAt(1));

            case Or:
                return checkModel(formula.getChildAt(0)) || checkModel(formula.getChildAt(1));

            case Implication:
                return !checkModel(formula.getChildAt(0)) || checkModel(formula.getChildAt(1));

            case Biimplication:
                boolean left  = checkModel(formula.getChildAt(0));
                boolean right = checkModel(formula.getChildAt(1));
                return (left && right) || (!left && !right);

            case Predicate:
                switch (formula.getChildren().size()) {
                    case 1:
                        return oneArySymbolTable.get(formula.getName()).contains(interpret(formula.getChildAt(0)));
                    case 2:
                        if (formula.getName().equals(FOLPredicate.INFIX_EQUALITY)) {
                            return interpret(formula.getChildAt(0)) == interpret(formula.getChildAt(1));
                        } else {
                            return twoArySymbolTable.get(formula.getName()).stream().anyMatch(edge ->
                             edge.getFromVertex().equals(interpret(formula.getChildAt(0))) &&
                                 edge.getToVertex().equals(interpret(formula.getChildAt(1)))
                            );
                        }
                    default:
                        throw new ModelCheckException("[ModelChecker][Internal error] Found predicate with to many children.");
                }

            case Constant:
                 return FOLConstant.TT.equals(formula.getName());

            default:
                throw new ModelCheckException("[ModelChecker][Internal error] Unknown FOLFormula-Type: " + formula.getType());
        }
    }

    /**
     * Takes a FOLFunction or FOLVariable and return the associated vertex within this interpretation.
     * @param symbol mus be a FOLFunction or an FOLVariable
     * @return a vertex which is associated with this term.
     * @throws ModelCheckException if the data is invalid. This should not happen.
     */
    private Vertex interpret(FOLFormula symbol) throws ModelCheckException {
        if (symbol instanceof FOLFunction) {
            if (symbol.getChildren().size() == 0) {
                return oneArySymbolTable.get(symbol.getName()).stream().findAny().get();
            } else if (symbol.getChildren().size() == 1) {
                Vertex childResult = interpret(symbol.getChildAt(0));
                return twoArySymbolTable.get(symbol.getName()).stream()
                    .filter(edge -> edge.getFromVertex() == childResult)
                    .findAny().get().getToVertex();
            } else {
                throw new ModelCheckException("[ModelChecker][Internal error] Found function with to many children.");
            }

        } else if (symbol instanceof FOLBoundVariable) {
            if(bindVariableValues.get(symbol.getName()) == null) {
                throw new ModelCheckException("[ModelChecker][Internal error] No bind value found for variable.");
            }
            return bindVariableValues.get(symbol.getName());
        } else {
            throw new ModelCheckException("[ModelChecker][Internal error] Not a valid function or a variable.");
        }
    }


    ////////////////////////// helping stuff ///////////////////////////////



    /**
     * A custom exception to handle error cases happened in the model check.
     */
    public static class ModelCheckException extends RuntimeException{
        /**
         * Constructs a new exception with the specified detail message.
         * @param message the detail message. The detail message is saved for
         *                later retrieval by the {@link #getMessage()} method.
         */
        public ModelCheckException(String message) {
            super(message);
        }
    }
}

File: src/eu/yeger/gramofo/fol/graph/ModelChecker.kt
package eu.yeger.gramofo.fol.graph

import com.github.michaelbull.result.*
import com.github.michaelbull.result.binding
import eu.yeger.gramofo.fol.Settings
import eu.yeger.gramofo.fol.formula.*
import eu.yeger.gramofo.fol.formula.FOLFormula.Companion.INFIX_EQUALITY
import eu.yeger.gramofo.model.api.TranslationDTO
import java.util.*
import kotlin.collections.MutableSet
import kotlin.collections.Set
import kotlin.collections.any
import kotlin.collections.component1
import kotlin.collections.component2
import kotlin.collections.forEach
import kotlin.collections.map
import kotlin.collections.mutableMapOf
import kotlin.collections.none
import kotlin.collections.set
import kotlin.collections.toSet

typealias ModelCheckerResult = Result<ModelCheckerTrace, TranslationDTO>

private val infixPredicates = Settings[Settings.INFIX_PRED].toSet()

private class ModelCheckException(message: String) : RuntimeException(message)

private data class SymbolTable(
    val unarySymbols: Map<String, Set<Node>>,
    val binarySymbols: Map<String, Set<Edge>>,
    val symbolTypes: Map<String, String>
)

fun checkModel(graph: Graph, formulaHead: FOLFormulaHead): ModelCheckerResult = binding {
    val symbolTable = graph.loadSymbols()
        .andThen { symbolTable -> formulaHead.loadSymbols(symbolTable) }
        .andThen { symbolTable -> checkTotality(graph, symbolTable) }.bind()
    runCatching { formulaHead.formula.checkFormula(graph, symbolTable, emptyMap()) }
        .mapError { error -> TranslationDTO(error.message ?: "api.error.unknown") }
        .bind()
}

private fun FOLFormula.checkFormula(graph: Graph, symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    return when (type) {
        FOLType.ForAll -> checkForAll(graph, symbolTable, variableAssignments)
        FOLType.Exists -> checkExists(graph, symbolTable, variableAssignments)
        FOLType.Not -> checkNot(graph, symbolTable, variableAssignments)
        FOLType.And -> checkAnd(graph, symbolTable, variableAssignments)
        FOLType.Or -> checkOr(graph, symbolTable, variableAssignments)
        FOLType.Implication -> checkImplication(graph, symbolTable, variableAssignments)
        FOLType.BiImplication -> checkBiImplication(graph, symbolTable, variableAssignments)
        FOLType.Predicate -> checkRelation(symbolTable, variableAssignments)
        FOLType.Constant -> checkConstant(variableAssignments)
        else -> throw ModelCheckException("[ModelChecker][Internal error] Unknown FOLFormula-Type: $type")
    }
}

private fun FOLFormula.checkForAll(graph: Graph, symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val variableName = getChildAt(0).name
    val children = graph.nodes.map { node: Node ->
        getChildAt(1).checkFormula(graph, symbolTable, variableAssignments + (variableName to node))
    }
    return if (children.all(ModelCheckerTrace::isModel)) {
        validated(TranslationDTO("api.forall.valid"), variableAssignments, *children.toTypedArray())
    } else {
        invalidated(TranslationDTO("api.forall.invalid"), variableAssignments, *children.toTypedArray())
    }
}

private fun FOLFormula.checkExists(graph: Graph, symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val variableName = getChildAt(0).name
    val children = graph.nodes.map { node: Node ->
        getChildAt(1).checkFormula(graph, symbolTable, variableAssignments + (variableName to node))
    }
    return if (children.any(ModelCheckerTrace::isModel)) {
        validated(TranslationDTO("api.exists.valid"), variableAssignments, *children.toTypedArray())
    } else {
        invalidated(TranslationDTO("api.exists.invalid"), variableAssignments, *children.toTypedArray())
    }
}

private fun FOLFormula.checkNot(graph: Graph, symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val child = getChildAt(0).checkFormula(graph, symbolTable, variableAssignments)
    return when (child.isModel) {
        true -> invalidated(TranslationDTO("api.not.invalid"), variableAssignments, child)
        false -> validated(TranslationDTO("api.not.valid"), variableAssignments, child)
    }
}

private fun FOLFormula.checkAnd(graph: Graph, symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val left = getChildAt(0).checkFormula(graph, symbolTable, variableAssignments)
    val right = getChildAt(1).checkFormula(graph, symbolTable, variableAssignments)
    return when {
        left.isModel && right.isModel -> validated(TranslationDTO("api.and.both"), variableAssignments, left, right)
        left.isModel.not() && right.isModel.not() -> invalidated(TranslationDTO("api.and.neither"), variableAssignments, left, right)
        left.isModel.not() -> invalidated(TranslationDTO("api.and.left"), variableAssignments, left, right)
        else -> invalidated(TranslationDTO("api.and.right"), variableAssignments, left, right)
    }
}

private fun FOLFormula.checkOr(graph: Graph, symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val left = getChildAt(0).checkFormula(graph, symbolTable, variableAssignments)
    val right = getChildAt(1).checkFormula(graph, symbolTable, variableAssignments)
    return when {
        left.isModel && right.isModel -> validated(TranslationDTO("api.or.both"), variableAssignments, left, right)
        left.isModel -> validated(TranslationDTO("api.or.left"), variableAssignments, left, right)
        right.isModel -> validated(TranslationDTO("api.or.right"), variableAssignments, left, right)
        else -> invalidated(TranslationDTO("api.or.neither"), variableAssignments, left, right)
    }
}

private fun FOLFormula.checkImplication(graph: Graph, symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val left = getChildAt(0).checkFormula(graph, symbolTable, variableAssignments)
    val right = getChildAt(1).checkFormula(graph, symbolTable, variableAssignments)
    return when {
        right.isModel -> validated(TranslationDTO("api.implication.right"), variableAssignments, left, right)
        left.isModel.not() -> validated(TranslationDTO("api.implication.left"), variableAssignments, left, right)
        else -> invalidated(TranslationDTO("api.implication.invalid"), variableAssignments, left, right)
    }
}

private fun FOLFormula.checkBiImplication(graph: Graph, symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val left = getChildAt(0).checkFormula(graph, symbolTable, variableAssignments)
    val right = getChildAt(1).checkFormula(graph, symbolTable, variableAssignments)
    return when (left.isModel) {
        right.isModel -> validated(TranslationDTO("api.bi-implication.valid"), variableAssignments, left, right)
        else -> invalidated(TranslationDTO("api.bi-implication.invalid"), variableAssignments, left, right)
    }
}

private fun FOLFormula.checkRelation(symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    return when (children.size) {
        1 -> checkUnaryRelation(symbolTable, variableAssignments)
        2 -> checkBinaryRelation(symbolTable, variableAssignments)
        else -> throw ModelCheckException("[ModelChecker][Internal error] Found predicate with to many children.")
    }
}

private fun FOLFormula.checkUnaryRelation(symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val node = getChildAt(0).interpret(symbolTable, variableAssignments)
    val translationParams = mapOf("relation" to name, "node" to node.name)
    return when (symbolTable.unarySymbols[name]!!.contains(node)) {
        true -> validated(TranslationDTO("api.relation.unary.valid", translationParams), variableAssignments)
        false -> invalidated(TranslationDTO("api.relation.unary.invalid", translationParams), variableAssignments)
    }
}

private fun FOLFormula.checkBinaryRelation(symbolTable: SymbolTable, variableAssignments: Map<String, Node>): ModelCheckerTrace {
    val left = getChildAt(0).interpret(symbolTable, variableAssignments)
    val right = getChildAt(1).interpret(symbolTable, variableAssignments)
    val translationParams = mapOf(
        "firstTerm" to getChildAt(0).toString(variableAssignments),
        "secondTerm" to getChildAt(1).toString(variableAssignments),
        "firstResult" to left.name,
        "secondResult" to right.name
    )
    return if (name == INFIX_EQUALITY) {
        when (left) {
            right -> validated(TranslationDTO("api.relation.equality.valid", translationParams), variableAssignments)
            else -> invalidated(TranslationDTO("api.relation.equality.invalid", translationParams), variableAssignments)
        }
    } else {
        val binaryTranslationParams = translationParams + ("relation" to name)
        when (symbolTable.binarySymbols[name]!!.any { edge: Edge -> edge.source == left && edge.target == right }) {
            true -> validated(TranslationDTO("api.relation.binary.valid", binaryTranslationParams), variableAssignments)
            else -> invalidated(TranslationDTO("api.relation.binary.invalid", binaryTranslationParams), variableAssignments)
        }
    }
}

private fun FOLFormula.checkConstant(variableAssignments: Map<String, Node>): ModelCheckerTrace {
    return when (name == FOLFormula.TT) {
        true -> validated(TranslationDTO("api.constant.true"), variableAssignments)
        else -> invalidated(TranslationDTO("api.constant.false"), variableAssignments)
    }
}

/**
 * Iterates over the graph and puts all found symbols in a symbol table.
 */
private fun Graph.loadSymbols(): Result<SymbolTable, TranslationDTO> {
    val unarySymbols = mutableMapOf<String, MutableSet<Node>>()
    val binarySymbols = mutableMapOf<String, MutableSet<Edge>>()
    val symbolTypes = mutableMapOf<String, String>()
    nodes.forEach { node: Node ->
        node.stringAttachments.forEach { symbol: String ->
            val symbolType = if (Character.isUpperCase(symbol[0])) "P-1" else "F-0"
            symbolTypes[symbol] = symbolType
            val relationSet = unarySymbols.getOrDefault(symbol, HashSet())
            if (symbolType == "F-0" && relationSet.size != 0) {
                return@loadSymbols Err(TranslationDTO(key = "api.error.duplicate-constant", mapOf("constant" to symbol)))
            }
            relationSet.add(node)
            unarySymbols[symbol] = relationSet
        }
    }
    edges.forEach { edge: Edge ->
        edge.stringAttachments.forEach { symbol: String ->
            val symbolType =
                if (Character.isUpperCase(symbol[0]) || infixPredicates.contains(symbol)) "P-2" else "F-1"
            symbolTypes.putIfAbsent(symbol, symbolType)
            if (symbolType != symbolTypes[symbol]) {
                return@loadSymbols Err(TranslationDTO("api.error.different-arities", mapOf("symbol" to symbol)))
            }
            val relationSet = binarySymbols.getOrDefault(symbol, HashSet())
            if (symbolType == "F-1" && relationSet.any { otherEdge: Edge -> edge.source == otherEdge.source }) {
                return@loadSymbols Err(TranslationDTO("api.error.different-function-values", mapOf("function" to symbol)))
            }
            relationSet.add(edge)
            binarySymbols[symbol] = relationSet
        }
    }
    return Ok(SymbolTable(unarySymbols = unarySymbols, binarySymbols = binarySymbols, symbolTypes = symbolTypes))
}

/**
 * Iterates over the formula and add new symbols to the symbol tables.
 */
private fun FOLFormulaHead.loadSymbols(symbolTable: SymbolTable): Result<SymbolTable, TranslationDTO> {
    val unarySymbols = symbolTable.unarySymbols.toMutableMap()
    val binarySymbols = symbolTable.binarySymbols.toMutableMap()
    val symbolTypes = symbolTable.symbolTypes.toMutableMap()
    this.symbolTable.forEach { (symbol: String, type: String) ->
        symbolTypes.putIfAbsent(symbol, type)
        if (type == "P-1" || type == "F-0") {
            unarySymbols.putIfAbsent(symbol, HashSet())
        } else if (type == "P-2" || type == "F-1") {
            binarySymbols.putIfAbsent(symbol, HashSet())
        }
        val typeInGraph = symbolTypes[symbol]
        if (type != typeInGraph) { // types are different?
            if (type == "V") {
                return@loadSymbols Err(TranslationDTO("api.error.bound-variable-reuse", mapOf("symbol" to symbol)))
            } else {
                return@loadSymbols Err(TranslationDTO("api.error.different-arities-formula", mapOf("symbol" to symbol)))
            }
        }
    }
    return Ok(SymbolTable(unarySymbols = unarySymbols, binarySymbols = binarySymbols, symbolTypes = symbolTypes))
}

/**
 * Functions mus be left total. Therefore this method checks if all function symbols are defined for all inputs.
 */
private fun checkTotality(graph: Graph, symbolTable: SymbolTable): Result<SymbolTable, TranslationDTO> {
    symbolTable.symbolTypes.forEach { (symbol: String, type: String) ->
        when (type) {
            "F-0" -> if (symbolTable.unarySymbols[symbol]!!.size != 1) {
                return@checkTotality Err(TranslationDTO("api.error.undefined-constant", mapOf("constant" to symbol)))
            }
            "F-1" -> {
                val relationSet: Set<Edge> = symbolTable.binarySymbols[symbol]!!
                graph.nodes.forEach { node: Node ->
                    if (relationSet.none { edge: Edge -> edge.source == node }) {
                        return@checkTotality Err(TranslationDTO("api.error.function-totality", mapOf("function" to symbol)))
                    }
                }
            }
        }
    }
    return Ok(symbolTable)
}

/**
 * Takes a FOLFunction or FOLVariable and return the associated node within this interpretation.
 * @return a node which is associated with this term.
 * @throws ModelCheckException if the data is invalid. This should not happen.
 */
@Throws(ModelCheckException::class)
private fun FOLFormula.interpret(symbolTable: SymbolTable, variableAssignments: Map<String, Node>): Node {
    return when (this) {
        is FOLFunction -> {
            when (children.size) {
                0 -> symbolTable.unarySymbols[name]!!.first()
                1 -> {
                    val childResult = getChildAt(0).interpret(symbolTable, variableAssignments)
                    symbolTable.binarySymbols[name]!!.first { edge: Edge -> edge.source == childResult }.target
                }
                else -> throw ModelCheckException("[ModelChecker][Internal error] Found function with to many children.")
            }
        }
        is FOLBoundVariable ->
            variableAssignments[name] ?: throw ModelCheckException("[ModelChecker][Internal error] Variable is not assigned.")
        else -> throw ModelCheckException("[ModelChecker][Internal error] Not a valid function or a variable.")
    }
}

private fun FOLFormula.validated(description: TranslationDTO, variableAssignments: Map<String, Node>, vararg children: ModelCheckerTrace) =
    ModelCheckerTrace(this.toString(variableAssignments), description, true, children.toList())

private fun FOLFormula.invalidated(description: TranslationDTO, variableAssignments: Map<String, Node>, vararg children: ModelCheckerTrace) =
    ModelCheckerTrace(this.toString(variableAssignments), description, false, children.toList())
